### ISSUE TITLE
[unbound-method] Should prevent conforming to an bound-style interface method with an unbound method

### ISSUE THREAD

**Issue by jtbandes** [typescript-eslint/typescript-eslint#3503]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
interface Foo {
  unbound(): number;
  unboundOk(this: void): number;
  bound: () => number;
}

function what(f: Foo) {
  const unbound = f.unbound;      // lint error 👍
  const unboundOk = f.unboundOk;  // no error 👍
  const bound = f.bound;          // no error 👍

  // console.log(unbound());  // runtime error (expected)
  console.log(unboundOk()); // runtime error ❌
  console.log(bound());     // runtime error ❌
}

class Example implements Foo {
  x = 3;
  unbound() { return this.x; }
  unboundOk() { return this.x; }   // no error 👎 (TypeScript issue rather than typescript-eslint issue?)
  bound() { return this.x; }      // no error 👎 (this issue)
}
what(new Example());
```

eslint config: https://github.com/foxglove/eslint-plugin/blob/main/configs/typescript.js
tsconfig: https://github.com/foxglove/studio/blob/main/packages/tsconfig/tsconfig.base.json

**Expected Result**

`class Example implements Foo` is problematic. There should be an error on `bound() { return this.x }` when is used to satisfy the interface requirement `bound: () => number;`.

**Actual Result**

The lint rule does not reject this incorrect method implementation, which leads to a runtime error.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.26.0` |
| `@typescript-eslint/parser`        | `4.26.0` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `15.9.0` |

**Comment by bradzacher** [comment]
It's important to note that you are misunderstanding what TS syntax means.
```ts
interface Foo {
  prop: () => number;
}
```

This does not mean or enforce that `prop` is an arrow function.
Whilst it looks similar to an arrow function property - **the two are not at all related**!

The difference between `prop: () => number` and `prop(): number` is only apparent when the `strictFunctionTypes` flag is turned on. You can read more in our [method-signature-style](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/method-signature-style.md) rule.
Otherwise though - a method signature and a property signature are just different syntax for the same thing.

The reason that `unbound-method` does not report on `bound` is simply because it does not have support in the rule for detecting "property signatures with function types".

**Comment by ronami** [comment]
I looked into this for a while, and have some thoughts/questions:

- What's the best way to resolve the issue?

  According to this issue (at least to my understanding), the rule should report on the class that wrongly implements an interface:

  ```ts
  interface Foo {
    bound: () => number;
  }
  
  class Example implements Foo {
    bound() { return this.x; }
  }
  ```
  
  Rather than report on the place where the unbound function is created:
  
  ```ts
  function what(f: Foo) {
    const bound = f.bound;
  }
  ```
  
- I think this is incorrect, as @bradzacher pointed out (and to my understand), there are small differences between the two ([see here](https://github.com/microsoft/TypeScript/pull/18654), also [described here](https://typescript-eslint.io/rules/method-signature-style/)), and I think both should be reported similarly - when the unbound function is created ([playground link](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIwCMAFAJQBcyIArgLYBG0A3PoQExXnIC8AfNXpMorAL65cAEwgIANnCgoE6EAGcwhTFQzpWuAPT7kyqIoRhZAT2SKADuiiRJuZWo3Fem9ADpiew8iqABboNLKSyEw2EPaOEM6u6hyeMJi%2B7MxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):
  
  ```ts
  interface Foo {
    f1(): number;
    f2: () => number;
  }
  
  declare const foo: Foo;

  // correctly reported
  const f1 = foo.f1;

  // should be reported
  const f2 = foo.f2;
  ```

  I don't think `f2` can be treated as a bound function, as it can easily end up with a missing or wrong `this` value (see this issue's repro for an example).

- As for reporting the class that wrong implements an interface, I think this is related more to #277, which should not only report that function parameters are assignable to the parent, but also its `this` argument ([playground link](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=ATD0uB3AnB7A7A5gGwJ7AKYA8AuH4AmAzsAIbABGpRGwAxstSRevKdHJAJZLAGyR4ZYADMArvDo4uCMtERiAtvhwAoEDzzQRpOrQAK7FQEZgAb3UhRACgAOALmDwlFDNGAAfYERzQeiAEpHADdYLgIAbksAX1VLBiZgAGEACy5kAlMuRVtkDGV4HBJDaBNzSxARO0cfPyQg4FDw82BYkFjLcCg4JDRMXHxiYSoaekYiZlZ2Tn8%2BASFycUlpWQADHDSiVbkFArUNQrcdPWASlQAmcqsbDa4iR2dFV3cvWv8GpsiYuJAEieS0hlLtlcvkVMUjIVLhZrlVbvdvL53iEwgQWm1Wj8wBBYIouDhpLxFhIpDJ4AByEjrTbbHCoWwYAA0UDSdBSwDu-VyXDo%2BL6tjgwXCGDRsg2tFskJw9ksmiOugMUoAzFcrHDNg8XG4Pqiou0sX8SKl0gQVSC8nsIaVCiqYWrrDrmnbrvCAHRVAJ6qwY2JAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=&tokens=false)), or potentially a different rule altogether. I'd love to hear more opinions on this.

- To my understanding, the best resolution to this would be to report `PropertySignature` similarly to `PropertyDeclaration`, and let #277 or a different rule to handle classes that wrongly extend an interface/class. I've pushed a draft PR to address this in https://github.com/typescript-eslint/typescript-eslint/pull/11682 and will be happy to open it for review if any of this makes sense.

- While working on the PR above, I noticed a similar issue, in which the rule only reports `PropertyDeclaration` with a function initializer, and ignores those with a function type (it does report `MethodDeclaration`, which is described with a type):
  
  ```ts
  class Base {
    bound?: () => void;
    unbound?(): void;
  }

  const instance = new Base();

  // reported correctly
  const unbound = instance.unbound;

  // should be reported but isn't?
  const bound = instance.bound;
  ```

  There seems to be [a test to make sure this isn't reported](https://github.com/typescript-eslint/typescript-eslint/blob/7f5fed7922385479613f4d79a28615014174a556/packages/eslint-plugin/tests/rules/unbound-method.test.ts#L21-L41), which I think is a mistake. See the following, for an error the rule fails to catch because of this ([playground link](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=MYGwhgzhAEBCkFNoG8BQ1oCMD2BXAdgCYD8AXNABQCU0AvAHzQBu2AloQNyrrTDb4QALgCdcwQdmHUUPDIIAWrCADocBQnWgAzAuNb9KNNBhO9%2BEbCATKQ2AOYUFS1XiJUupgL49PGVN9Q%2BAUFoVmCwfGAkWmh8BAB3OERqLkDzELUiTTChCKiXdVSAeiLoQCTAQxQzAUtrWwcnFUzCd2hvHME862biZRTuEugAFQBPAAcEAFFhYUlyAGEI-GwQ4QQwDTHZieFBVgQYbC1odQQtMIQNCjWNsLtoAHJmh6pUHr73VCA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):
  
  ```ts
  class Base {
    bound?: () => void;

    constructor() {
      this.bound = function () {
        console.log(this.bound);
     }
    }  
  }

  const instance = new Base();

  const bound = instance.bound;

  // ƒ () { console.log(this.bound); }
  instance.bound?.();

  // TypeError: Cannot read properties of undefined (reading 'bound')
  bound?.();
  ```

  Since the rule reports an object with a function property, I think it should do the same for a class ([playground link](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=MYewdgzgLgBCBGArGBeGBvAUDGAzEIAXHgK5jBQCW4AFAJQYC%2BmzmA9GzAE4CmADiC5QeAExiguvCgBsAnplCRY%2BEKjhIAdCoDcmIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDToMmeCmFHopAXxAqgA&tokens=false)):
  
  ```ts
  const obj = {
    foo: function() {}
  }
  
  // reported correctly
  const foo = obj.foo;
  ```

  Please let me know if this makes sense (I can open a separate issue to discuss this), I'm not sure if I'm missing an important edge case or not.

**Comment by bradzacher** [comment]
So we ignore the arrow function propert signatures on types because (IIRC) it's just something that was carried over from the previous tslint implementation.

It was definitely an incorrect thing but it's also a bit of a DevX concession IIRC cos people were using the property signature style to annotate a bound method. I also think that types wise TS also generates a type with a property signature for arrow function properties.

Ultimately there's no amazing solution for this. The rule is an estimation at best.

**Comment by bradzacher** [comment]
Based on our current expectations for enhancement requests, I believe this request fails the "nicheness" test:
- This request has been open for 5 years and has netted a total of 1 non-team member interactions.
- This request has been open for 5 years and nobody has moved to implement it.

For these reasons, I believe we should move to reject and close this. Thoughts @typescript-eslint/triage-team?

**Comment by JoshuaKGoldberg** [comment]
:c I think it'd be nice and valuable. This is a tricky area not covered by TS itself.
